qiskit/circuit/random/utils.py

```diff
@@ -11,15 +11,394 @@
 import numpy as np
 
-from qiskit.circuit import ClassicalRegister, QuantumCircuit, CircuitInstruction
+from qiskit.circuit import (
+    ClassicalRegister,
+    QuantumCircuit,
+    CircuitInstruction,
+    Qubit,
+    QuantumRegister,
+)
 from qiskit.circuit import Reset
 from qiskit.circuit.library import standard_gates
 from qiskit.circuit.exceptions import CircuitError
 
 
+def _get_gates():
+    gates_1q = [
+        # (Gate class, number of qubits, number of parameters)
+        (standard_gates.IGate, 1, 0),
+        (standard_gates.SXGate, 1, 0),
+        (standard_gates.XGate, 1, 0),
+        (standard_gates.RZGate, 1, 1),
+        (standard_gates.RGate, 1, 2),
+        (standard_gates.HGate, 1, 0),
+        (standard_gates.PhaseGate, 1, 1),
+        (standard_gates.RXGate, 1, 1),
+        (standard_gates.RYGate, 1, 1),
+        (standard_gates.SGate, 1, 0),
+        (standard_gates.SdgGate, 1, 0),
+        (standard_gates.SXdgGate, 1, 0),
+        (standard_gates.TGate, 1, 0),
+        (standard_gates.TdgGate, 1, 0),
+        (standard_gates.UGate, 1, 3),
+        (standard_gates.U1Gate, 1, 1),
+        (standard_gates.U2Gate, 1, 2),
+        (standard_gates.U3Gate, 1, 3),
+        (standard_gates.YGate, 1, 0),
+        (standard_gates.ZGate, 1, 0),
+    ]
+
+    gates_2q = [
+        (standard_gates.CXGate, 2, 0),
+        (standard_gates.DCXGate, 2, 0),
+        (standard_gates.CHGate, 2, 0),
+        (standard_gates.CPhaseGate, 2, 1),
+        (standard_gates.CRXGate, 2, 1),
+        (standard_gates.CRYGate, 2, 1),
+        (standard_gates.CRZGate, 2, 1),
+        (standard_gates.CSXGate, 2, 0),
+        (standard_gates.CUGate, 2, 4),
+        (standard_gates.CU1Gate, 2, 1),
+        (standard_gates.CU3Gate, 2, 3),
+        (standard_gates.CYGate, 2, 0),
+        (standard_gates.CZGate, 2, 0),
+        (standard_gates.RXXGate, 2, 1),
+        (standard_gates.RYYGate, 2, 1),
+        (standard_gates.RZZGate, 2, 1),
+        (standard_gates.RZXGate, 2, 1),
+        (standard_gates.XXMinusYYGate, 2, 2),
+        (standard_gates.XXPlusYYGate, 2, 2),
+        (standard_gates.ECRGate, 2, 0),
+        (standard_gates.CSGate, 2, 0),
+        (standard_gates.CSdgGate, 2, 0),
+        (standard_gates.SwapGate, 2, 0),
+        (standard_gates.iSwapGate, 2, 0),
+    ]
+    gates_3q = [
+        (standard_gates.CCXGate, 3, 0),
+        (standard_gates.CSwapGate, 3, 0),
+        (standard_gates.CCZGate, 3, 0),
+        (standard_gates.RCCXGate, 3, 0),
+    ]
+    gates_4q = [
+        (standard_gates.C3SXGate, 4, 0),
+        (standard_gates.RC3XGate, 4, 0),
+    ]
+    return (gates_1q, gates_2q, gates_3q, gates_4q)
+
+
+def random_circuit_from_graph(
+    interaction_graph,
+    min_2q_gate_per_edge,
+    max_operands=2,
+    measure=False,
+    conditional=False,
+    reset=False,
+    seed=None,
+    insert_1q_oper=False,
+    prob_conditional: float = 0.1,
+):
+    """Generate random circuit of arbitrary size and form which strictly respects
```

Comment:
It also seems like there could be a significant variance in circuit depth for graphs that have very varying edge weights. How about using the total number of gates as a standard termination criterion? Otherwise it would be important to clearly point out what the termination criterion is 🙂

Reply:
@Cryoris , 
Couldn't agree more :) 

In that spirit, I have added an argument `rel_diff` read the docstring for more information.

Reply:
![rel_diff](https://github.com/Qiskit/qiskit/assets/54737489/f3563f68-b4b9-4c1f-93a7-36939e601517)

Reply:
Although its not a linear knob, but at-least this can be used to keep check on the size of the circuit generated.

I know these generated circuits are supposed to be fed to transpiler running on Rust,  which is really fast.
But, in case if a user wants to draw the circuits, this would be a lot of overhead, if the number of gates, especially measure is more.

Let me know what the reviewer thinks, I have just added the code for `rel_diff` and not added the docstring and not mentioned in the release note. I will add or remove those on the sayings of the reviewer.

Reply:
How `rel_diff` works,

Post normalization of the edge weights, I have taken a relative difference between the minimum and maximum probabilities, because this is the variable that would determine the depth of the circuit.

A user can keep a check on the circuit size, by controlling this variable `rel_diff`.

Reply:
I think it's better to keep the function as simple as possible, especially on last minute changes. It might be best to remove `rel_diff` and just clearly state how this function chooses the termination 🙂